src/hashbang.sh

```diff
@@ -156,99 +166,158 @@ echo " over the internet, and thus by nature we won't even know what it is";
 echo " ";
 
 for keytype in id_rsa id_dsa id_ecdsa id_ed25519; do
-    if [ -e ~/.ssh/$keytype.pub  ]; then
-        if ask " We found a public key in [ ~/.ssh/$keytype.pub ]. Use this key?" Y; then
-            keyfile="~/.ssh/$keytype.pub"
-            key=$(cat ~/.ssh/$keytype.pub)
-            break
-        fi
-    fi
+	if [ -e ~/.ssh/$keytype.pub ]; then
+		if ask " We found a public key in [ ~/.ssh/$keytype.pub ]. Use this key?" Y; then
+			keyfile="~/.ssh/$keytype.pub"
+			key=$(cat ~/.ssh/$keytype.pub)
+			break
+		fi
+	fi
 done
 
+makekey() {
+	( checkutil ssh-keygen && checkutil ssh ) || exit 1
+	if checkperms "$1"; then
+		ssh-keygen -t rsa -C "#! $username" -f "$1"
+		if [ ! $? ]; then
+			echo " Unable to make key with that location"
+			echo " "
+		else
+			echo " Successfully generated key"
+			echo " "
+			break
+		fi
+	else
+		if [ -e "$1" ]; then
+			if ask " Unable to generate key, do you want to delete the file?" N; then
+				rm -f "$1"
+				if [ ! $? ]; then
+					echo " "
+					echo " Unable to delete file, resetting"
+				else
+					echo " "
+					echo " File deleted"
+					ssh-keygen -t rsa -C "#! $username" -f "$1"
+					if [ ! $? ]; then
+						echo " Unable to generate key, resetting"
+					fi
+					echo " "
+				fi
+			else
+				echo " "
+				echo " Unable to make key with that path, resetting"
+				echo " "
+			fi
+		else
+			echo "Unable to make key with that path"
+			echo " "
+		fi
+	fi
+}
+
 if [ "x$key" = "x" ]; then
-    if ask " Do you want us to generate a key for you?" Y; then
-        ssh-keygen -t rsa -C "#! $username"
-        keyfile="~/.ssh/id_rsa.pub"
-        key=$(cat ~/.ssh/id_rsa.pub)
-    fi
+	while true; do
+		echo " "
+		echo -n " Path to file (~/.ssh/id_rsa): ";
```

Comment:
Many users have no clue what public/private keys are. Here it would probably be better to do:

Would you like us to create a key at the default location ( ~/.ssh/id_rsa.pub) ?

If yes, (either they are fine with this, or don't know what it means) then we skip the read and proceed with ~/.ssh/id_rsa.pub, else they know what they are doing, and want to specify something custom, so we let them.